### PR TITLE
fix(commonjs): add .cjs to default file extensions.

### PR DESCRIPTION
**Pull request by christiantjl**
## Rollup Plugin Name: `commonjs`

This PR contains:

- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

 

List any relevant issue numbers: #523 

### Description

Adds `.cjs` to default file extensions.

### CODE REVIEW

File: packages/commonjs/yarn.lock [hunk 1]
```diff
@@ -0,0 +1,287 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (shellscape) [inline packages/commonjs/yarn.lock:1]: please remove this file from the commit. this repo does not use `yarn`

**Comment by lukastaegert** [comment]
While I very much agree that we need to add `.cjs`, what was your reasoning of adding `.mjs` as well? After all by their definition, they cannot contain CommonJS elements.

**Comment by christiantjl** [comment]
@lukastaegert You're right, that definitely doesn't make sense - will update.

File: packages/commonjs/src/index.js [hunk 2]
```diff
@@ -34,7 +34,7 @@ import {
 } from './transform';
 
 export default function commonjs(options = {}) {
-  const extensions = options.extensions || ['.js'];
+  const extensions = options.extensions || ['.js', '.cjs'];
```
Comment (Andarist) [inline packages/commonjs/src/index.js:37]: q: what exact scenario has been failing and why? from what I understand `.cjs` (nor `.mjs`) can ever be implicit (please correct me if I'm wrong), so what `@rollup/plugin-commonjs` has been doing that it didn't handle this correctly? 🤔

Comment (Andarist) [reply packages/commonjs/src/index.js:37]: Looking at `getCandidates` here:
https://github.com/rollup/plugins/blob/8c49103deb620b26d0d019ee19c3d1420a8933dc/packages/commonjs/src/resolve-id.js#L23
it seems that explicit `.cjs` should always work as it's most likely the `resolved` variable here unless I miss something.

Comment (christiantjl) [reply packages/commonjs/src/index.js:37]: I included .mjs erroneously, but this PR was prompted by the behavior in #523, where the output would have a `module.exports` declaration at the top as well as ES2015 `export`s at the bottom when rolling in a .cjs file.

Comment (lukastaegert) [reply packages/commonjs/src/index.js:37]: Ah, I think the problem is that those extensions have a double meaning:
1. They allow for implicit resolution
2. They are being used as a filter: https://github.com/rollup/plugins/blob/5fa15590e452d568239a6d8e1c7018865c027fe4/packages/commonjs/src/index.js#L192

The latter is what the problem is here.

Comment (lukastaegert) [reply packages/commonjs/src/index.js:37]: But you are right, `.cjs` should not be allowed for implicit resolution by default as this contradicts the algorithm in Node.

Comment (shellscape) [reply packages/commonjs/src/index.js:37]: @christiantjl please review these recent comments

Comment (christiantjl) [reply packages/commonjs/src/index.js:37]: @shellscape I'm pretty new to Rollup and Node in general, could you dumb down these comments for me? Is the issue that my change will extend to like, `require('./dir/root')` catching `./dir/root.cjs` when that would conflict with Node logic? Any advice on remedying?

Comment (Andarist) [reply packages/commonjs/src/index.js:37]: > Is the issue that my change will extend to like, require('./dir/root') catching ./dir/root.cjs when that would conflict with Node logic?

Yes, this should not resolve by default to `.cjs`. 

> Any advice on remedying?

The applied change is not correct - as @lukastaegert has observed it seems that `extensions` are used now by 2 separate places in the logic. You have added `.cjs` to the first one incorrectly (allowing for implicit resolution) but actually `.cjs` should just be passed through this condition at all times:
https://github.com/rollup/plugins/blob/5fa15590e452d568239a6d8e1c7018865c027fe4/packages/commonjs/src/index.js#L191-L196
As we know that `.cjs` has to be in the commonjs format.

Comment (christiantjl) [reply packages/commonjs/src/index.js:37]: OK, thank you very much for explaining @Andarist. I will update shortly.

**Pull request merged**